draft-duke-quic-retry-offload.md

```diff
@@ -0,0 +1,876 @@
+
+Retry tokens generated by the offload MUST have the format below.
+
+~~~
+No-Shared-State Retry Offload Token {
+  Token Type (1) = 0,
+  ODCIL (7) = 8..20,
+  Original Destination Connection ID (64..160),
+  Opaque Data (..),
+}
+~~~
+{: #nss-retry-offload-token-format title="Format of non-shared-state Retry Offload tokens"}
+
+The first bit of retry tokens generated by the offload MUST be zero. The token
+has the following additional fields:
+
+ODCIL: The length of the original destination connection ID from the triggering
+Initial packet. This is in cleartext to be readable for the server, but
+authenticated later in the token. The Retry Offload SHOULD reject any token
+in which the value is less than 8.
+
+Original Destination Connection ID: This also in cleartext and authenticated
+later.
+
+Opaque Data: This data contains the information necessary to authenticate the
+Retry token in accordance with the QUIC specification. A straightforward
+implementation would encode the Retry Source Connection ID, client IP address,
+and a timestamp in the Opaque Data. A more space-efficient implementation would
+use the Retry Source Connection ID and Client IP as associated data in an
+encryption operation, and encode only the timestamp and the authentication tag
+in the Opaque Data. If the Initial Packet has altered the Connection ID or
+source IP address, authentication of the token will fail.
+
+Upon receipt of an Initial packet with a token that begins with '0', the retry
+offload MUST validate the token in accordance with the QUIC specification.
+
+In active mode, the offload MUST issue Retry packets for all Client initial
+packets that contain no token, or a token that has the first bit set to '1'. It
+MUST NOT forward the packet to the server. The offload MUST validate all tokens
+with the first bit set to '0'. If successful, the offload MUST forward the
+packet with the token intact. If unsuccessful, it MUST drop the packet. The
+Retry Offload MAY send an Initial Packet containing a CONNECTION_CLOSE frame
+with the INVALID_TOKEN error code when dropping the packet.
+
+Note that this scheme has a performance drawback. When the Retry Offload is in
+active mode, clients with a token from a NEW_TOKEN frame will suffer a 1-RTT
+penalty even though its token provides proof of address.
+
+In inactive mode, the offload MUST forward all packets that have no token or a
+token with the first bit set to '1'. It MUST validate all tokens with the first
+bit set to '0'. If successful, the offload MUST forward the packet with the
+token intact. If unsuccessful, it MUST either drop the packet or forward it
+with the token removed. The latter requires decryption and re-encryption of the
+entire Initial packet to avoid authentication failure. Forwarding the packet
+causes the server to respond without the original_destination_connection_id
+transport parameter, which preserves the normal QUIC signal to the client that
+there is an on-path attacker.
```

Comment:
While I agree that in an ideal world informing the client of an on-path attacker might be nice, I don't think any server would waste the resources to decrypt and re-encrypt, and I don't think it's even worth calling this out as an option. I think we should simply recommend drop.

Reply:
I think that's fine, but please file an issue.